zcash_client_backend/src/data_api/wallet.rs

```diff
@@ -772,104 +795,108 @@ where
+    let mut metadata_from_address = |addr: TransparentAddress| -> Result<
+        TransparentAddressMetadata,
+        ErrorT<DbT, InputsErrT, FeeRuleT>,
+    > {
+        match known_addrs.get(&addr) {
+            None if !ephemeral_added => {
+                // The ephemeral addresses are added lazily to avoid extra database operations
+                // in the common case. We don't need to include them in order to be able to
+                // construct ZIP 320 transactions, because in that case the ephemeral output
+                // is represented via a "change" reference to a previous step. However, we do
+                // need them in order to create a transaction from a proposal that explicitly
+                // spends an output from an ephemeral address. This need not set `for_detection`
+                // because we only need to be able to spend outputs already detected by this
+                // wallet instance.
+                ephemeral_added = true;
+                known_addrs.extend(
+                    wallet_db
+                        .get_reserved_ephemeral_addresses(account_id, false)
+                        .map_err(Error::DataSource)?
+                        .into_iter(),
+                );
+                known_addrs.get(&addr)
+            }
+            result => result,
+        }
+        .ok_or(Error::AddressNotRecognized(addr))?
+        .clone()
+        .ok_or_else(|| Error::NoSpendingKey(addr.encode(params)))
+    };
 
+    #[cfg(feature = "transparent-inputs")]
+    let utxos_spent = {
         let mut utxos_spent: Vec<OutPoint> = vec![];
-        let mut add_transparent_input = |addr: &TransparentAddress,
+        let mut add_transparent_input = |address_metadata: &TransparentAddressMetadata,
                                          outpoint: OutPoint,
-                                         utxo: TxOut|
-         -> Result<
-            (),
-            Error<
-                <DbT as WalletRead>::Error,
-                <DbT as WalletCommitmentTrees>::Error,
-                InputsErrT,
-                FeeRuleT::Error,
-            >,
-        > {
-            let address_metadata = known_addrs
-                .get(addr)
-                .ok_or(Error::AddressNotRecognized(*addr))?
-                .clone()
-                .ok_or_else(|| Error::NoSpendingKey(addr.encode(params)))?;
-
+                                         txout: TxOut|
+         -> Result<(), ErrorT<DbT, InputsErrT, FeeRuleT>> {
             let secret_key = usk
                 .transparent()
                 .derive_secret_key(address_metadata.scope(), address_metadata.address_index())
-                .unwrap();
+                .expect("spending key derivation should not fail");
 
             utxos_spent.push(outpoint.clone());
-            builder.add_transparent_input(secret_key, outpoint, utxo)?;
+            builder.add_transparent_input(secret_key, outpoint, txout)?;
```

Comment:
Isn't this by definition necessarily a utxo?

Reply:
I renamed it because the type is `TxOut`. Strictly speaking we don't know that it is a UTXO *here*, even though the spend will only succeed if it is unspent.